src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,243 @@
+
+//! This file contains the Graham algorithm to incrementally create or maintain a convex hull
+//! as well as the necessary structures, enums, and auxiliary functions. The algorithm is described
+//! in: "Mark De Berg, and Marc van Kreveld. Computational geometry: algorithms and applications.
+//! Springer Science & Business Media, 2000. https://doi.org//10.1007/978-3-540-77974-2.
+//! Convex Hulls are necessary in the implementation of the compression algorithms [1] and [2].
+//! [1] https://doi.org/10.14778/1687627.1687645 (Slide Filter).
+//! [2] https://doi.org/10.1109/TSP.2006.875394 (Optimal PLA).
+
+const std = @import("std");
+const mem = std.mem;
+const testing = std.testing;
+
+const tersets = @import("../../tersets.zig");
+const Error = tersets.Error;
+
+/// Enum for the angle's `Turn` of three consecutive points. The angle can represent a `right` or
+/// `left` turn. If there is no turn, then the points are `colinear`.
+const Turn = enum(i8) { right, left, colinear };
+
+/// Point structure to represent a point by `time` and `value`.
+const Point = struct { time: usize, value: f64 };
+
+/// Set of points used to store the upper and lower hull of the convex hull. `points` are memory
+/// slide that it's dynamically increased as `len` exceeds `max_len`. A good estimation of
+/// `max_len` can improve both execution time and memory consumption.
+pub const PointSet = struct {
+    points: []Point,
+    len: usize,
+    max_len: usize,
+    allocator: *const mem.Allocator,
+
+    // Initialize the container with a given `allocator` and max number of points `num_points`.
+    // The max number of points it is not fixed. The max number of points will increase as more
+    // elements are added.
+    pub fn init(allocator: *const mem.Allocator, num_points: usize) !PointSet {
+        return PointSet{
+            .points = try allocator.alloc(Point, num_points),
+            .len = 0,
+            .max_len = num_points,
+            .allocator = allocator,
+        };
+    }
+
+    // Deinitialize the container and free the allocated memory.
+    pub fn deinit(self: *PointSet) void {
+        self.allocator.free(self.points);
+    }
+
+    // Add `point` to the `PointSet`.
+    pub fn add(self: *PointSet, point: Point) !void {
+        if (self.len >= self.max_len) {
+            // Resize the set of `points`.
+            self.max_len = self.max_len * 2;
+            self.points = try self.allocator.realloc(self.points, self.max_len);
+        }
+        self.points[self.len] = point;
+        self.len += 1;
+    }
+
+    // Remove the last point from the set.
+    pub fn pop(self: *PointSet) !void {
+        if (self.len == 0) return Error.EmptySet;
+        self.len -= 1;
+    }
+};
+
+/// Graham algorithm to add a new point to `upper_hull` and `lower_hull`. The algorithm ensures
+/// that the upper and lower hull form a Convex Hull.
+pub fn addPointToConvexHull(upper_hull: *PointSet, lower_hull: *PointSet, point: Point) !void {
```

Comment:
It seems unnecessarily complicated for the user to manage the convex hull as two separate `PointSet`s. Maybe create a type named `ConvexHull` that internally stores `upper_hull` and `lower_hull` and implements needed for operating on the convex hull. This would also more directly model the "domain" and thus hopefully be simpler to understand for the user then two half hulls. If this change is made the files doc comment should be updated to reflect it.

Reply:
Fixed. I have changed the structure to ConvexHull storing the upper and lower hulls as ArrayList(Point). I have changed all documentation accordingly.